### ISSUE TITLE
UID null pointer crash

### ISSUE THREAD

**Issue by acepho** [fluffos/fluffos#459]
VER 3.0.20170907 Segmentation fault crash.

It's confirmed that crash happens when an object has null UID.

in driver source, packages/uids/uids.cc
```
void f_getuid(void) {
  ob = sp->u.ob;

  DEBUG_CHECK(ob->uid == NULL, "UID is a null pointer\n");
  put_constant_string(ob->uid->name); // !!! crashed here !!!
  free_object(&ob, "f_getuid");
}
`

There was a "DEBUG_CHECK" at the first place, My guess it happened before. Maybe worth reporting back.

I don't know how to repeat it. Don't know how to make an object uid == null. Also, I'd like to know if my mudlib code had flaws. So I added the following line in uids.cc.

`if (ob->uid == NULL) error("object uid is null!\n");`

It worked and I could trace it further.

##### Bug Log #####
执行时段错误：*object uid is null!

程式：/adm/core/simul_efun.c 第 77 行
物件: /adm/core/simul_efun
呼叫来自：/adm/core/master.c 的 connect() 第 13 行，物件： /adm/core/master
呼叫来自：/adm/core/master.c 的 CATCH() 第 13 行，物件： /adm/core/master
呼叫来自：/adm/core/master.c 的 creator_file() 第 209 行，物件： /adm/core/master
呼叫来自：/adm/core/simul_efun.c 的 creator_file() 第 77 行，物件： /adm/core/simul_efun
##### End of Log #####

mudlib simul_efun.c
`
string creator_file(string file) // this is called by driver, I don't know exactly when.
{
...
			if (this_player(1)) return getuid(this_player(1)); // !!! crashed here !!!
			else return "MudOS";
...
}
`

backward a bit mudlib master.c
`
object connect()
{
	object login_ob;
	mixed err;
	err = catch(login_ob = new(LOGIN_OB)); // !!! crashed here !!!
	if (err)
	{
	}
	return login_ob;
}
`

In this case, it seems when I tried to login to my mud, the line in master.c 'err = catch(login_ob = new(LOGIN_OB));' triggered the driver to call "creator_file()" in simul_efun, while "this_player(1)" had a null UID. At the moment no-one was online in the mud. I login to this mud from time to time, only triggered this once in many days. So this is strange. Don't know what is "this_player(1)" in such situation. Will add code in my mudlib to log it.

This is as far as I could get. Will report back if there is new finding.


##### Early log #####
error: SIGSEGV: Segmentation fault

----- Core dump -----
Core was generated by `./driver sys.cfg -d'.
Program terminated with signal 6, Aborted.
#0  0x00007f6142b14207 in raise () from /lib64/libc.so.6
Missing separate debuginfos, use: debuginfo-install glibc-2.17-260.el7_6.4.x86_64 libevent-2.0.21-4.el7.x86_64 libgcc-4.8.5-36.el7_6.2.x86_64 libstdc++-4.8.5-36.el7_6.2.x86_64 nss-softokn-freebl-3.36.0-5.el7_5.x86_64 pcre-8.32-17.el7.x86_64 zlib-1.2.7-18.el7.x86_64
(gdb) backtrace all
No symbol "all" in current context.
(gdb) backtrace full
#0  0x00007f6142b14207 in raise () from /lib64/libc.so.6
No symbol table info available.
#1  0x00007f6142b158f8 in abort () from /lib64/libc.so.6
No symbol table info available.
#2  0x0000559f14e06424 in fatal (fmt=<optimized out>) at vm/internal/simulate.cc:1692
        in_fatal = 0
        msg_buf = "SIGSEGV: Segmentation fault\000a\177\000\000萪_Da\177\000\000b\324>Da\177\000\000\000\000\000\000a\177\000\000繯闎a\177\000\000\377\377\377\377\000\000\000\000?\000\000\000\000\000\000\000\001\000\000\000\002\000\000\000`\b闎a\177\000\000繯闎a\177\000\000峤睟a\177\000\000\200ルBa\177\000\000繯闎a\177\000\000\000\000\000\000\071-05-18 20:43:51] comm.cc:1161 Closing connection from 193.188.22.69:6485.\n\000淿艬a"...
        args = {{gp_offset = 8, fp_offset = 48, overflow_arg_area = 0x7ffc53da2820, reg_save_area = 0x7ffc53da2750}}
        in_fatal = 0
#3  0x0000559f14de8a12 in (anonymous namespace)::attempt_shutdown(int) [clone .39459] (sig=<optimized out>) at mainlib.cc:182
        msg = 0x559f14e8935e "SIGSEGV: Segmentation fault"
#4  <signal handler called>
No symbol table info available.
#5  0x0000559f14d865b3 in f_getuid() () at packages/uids/uids.cc:68
No locals.
#6  0x0000559f14d44252 in eval_instruction(char*) (p=<optimized out>) at vm/internal/base/interpret.cc:3625
        i = 0
        real = 6.952576515114777e-310
        n = <optimized out>
        lval = <optimized out>
        instruction = 397
        real_instruction = 397
        offset = 65535
        msg = "with value: 0"
#7  0x0000559f14da826a in apply_low (fun=fun@entry=0x7f61422a8c08 "creator_file", ob=<optimized out>, ob@entry=0x7f61422830e0, num_arg=num_arg@entry=1) at vm/internal/apply.cc:276
        need = <optimized out>
        findex = 22
        funflags = <optimized out>
        runtime_index = <optimized out>
        funp = 0x7f61423529f8
        entry = {progp = 0x7f6142352000, funp = 0x7f61423529f8, function_index_offset = 0, variable_index_offset = 0}
#8  0x0000559f14da8426 in apply (fun=0x7f61422a8c08 "creator_file", ob=0x7f61422830e0, num_arg=1, where=<optimized out>) at vm/internal/apply.cc:313
No locals.
#9  0x0000559f14dbe4ab in f__call_other () at packages/core/efuns_main.cc:233
        num_arg = 3
        arg = 0x559f1510c1e0 <_ZL14start_of_stack.13354+64>
        funcname = 0x7f61422a8c08 "creator_file"
        ob = 0x7f61422830e0
#10 0x0000559f14d474e5 in eval_instruction(char*) (p=<optimized out>) at vm/internal/base/interpret.cc:3651
        num = 2
        i = 3
        real = 6.9196160747586565e-310
        n = <optimized out>
        lval = <optimized out>
        instruction = 134
        real_instruction = 134
        offset = 0
        msg = "with value: 0"
#11 0x0000559f14e01b15 in call_direct (num_arg=1, origin=1, offset=1, ob=0x7f61422831c0) at vm/internal/base/interpret.cc:3996
        funp = <optimized out>
        prog = 0x7f6142362000
#12 apply_master_ob (fun=<optimized out>, num_arg=1) at vm/internal/master.cc:37
No locals.
#13 0x0000559f14db02cd in give_uid_to_object(object_t*) [clone .5945] (ob=0x7f614228e400) at vm/internal/simulate.cc:177
        ret = <optimized out>
        creator_name = <optimized out>
#14 0x0000559f14dce5be in init_object (ob=0x7f614228e400) at vm/internal/simulate.cc:257
No locals.
#15 clone_object (str1=<optimized out>, num_arg=0) at vm/internal/simulate.cc:645
        ob = <optimized out>
        new_ob = 0x7f614228e400
#16 0x0000559f14dcea7b in f__new () at packages/core/efuns_main.cc:421
        ob = <optimized out>
#17 0x0000559f14d474e5 in eval_instruction(char*) (p=p@entry=0x7f61423620b2 "\016\003", <incomplete sequence \213>) at vm/internal/base/interpret.cc:3651
        num = 1
        i = 2
        real = 0
        n = <optimized out>
        lval = <optimized out>
        instruction = 139
        real_instruction = 139
        offset = 305
        msg = "with value: 0"
#18 0x0000559f14d45638 in do_catch (new_pc_offset=19, pc=0x7f61423620b2 "\016\003", <incomplete sequence \213>) at vm/internal/base/interpret.cc:3708
        econ = {save_csp = 0x559f150d4260 <control_stack>, save_sp = 0x559f1510c1b0 <_ZL14start_of_stack.13354+16>, save_cgsp = 0x559f150d2bc0 <_ZL19command_giver_stack.17143.6120>, save_context = 0x7ffc53da4ec0}
#19 eval_instruction(char*) (p=<optimized out>) at vm/internal/base/interpret.cc:3564
        i = 0
        real = 1.5810100666919889e-322
        n = <optimized out>
        lval = <optimized out>
        instruction = 39
        real_instruction = 39
        offset = 19
        msg = "with value: 0"
#20 0x0000559f14d8288a in apply_low (num_arg=1, ob=<optimized out>, fun=<optimized out>) at vm/internal/apply.cc:276
        need = <optimized out>
        findex = 15
        funflags = <optimized out>
        runtime_index = <optimized out>
        funp = 0x7f61423624c8
        local_call_origin = <optimized out>
        entry = {progp = 0x7f6142362000, funp = 0x7f61423624c8, function_index_offset = 0, variable_index_offset = 0}
#21 apply (fun=<optimized out>, ob=<optimized out>, where=where@entry=1, num_arg=1) at vm/internal/apply.cc:313
No locals.
#22 0x0000559f14d85676 in safe_apply (fun=<optimized out>, ob=<optimized out>, where=1, num_arg=1) at vm/internal/apply.cc:342
        econ = {save_csp = 0x559f150d4218, save_sp = 0x559f1510c1a0 <_ZL14start_of_stack.13354>, save_cgsp = 0x559f150d2bc0 <_ZL19command_giver_stack.17143.6120>, save_context = 0x0}
        ret = 0x0
#23 0x0000559f14db9d87 in safe_apply_master_ob (num_arg=1, fun=2) at vm/internal/master.cc:53
---Type <return> to continue, or q <return> to quit---
No locals.
#24 (anonymous namespace)::new_user_handler(evconnlistener*, int, sockaddr*, int, void*) [clone .37936] (listener=<optimized out>, fd=<optimized out>, addr=<optimized out>, addrlen=<optimized out>, arg=<optimized out>) at comm.cc:232
        master = 0x7f61422831c0
        ret = <optimized out>
        port = <optimized out>
        user = 0x7f61420e4800
        ob = <optimized out>
#25 0x00007f614390301a in listener_read_cb () from /lib64/libevent-2.0.so.5
No symbol table info available.
#26 0x00007f61438f6a14 in event_base_loop () from /lib64/libevent-2.0.so.5
No symbol table info available.
#27 0x0000559f14deff91 in backend(event_base*) (base=0x7f6142229000) at backend.cc:219
        t = {tv_sec = 1, tv_usec = 0}
#28 0x0000559f14d32052 in main (argc=<optimized out>, argv=<optimized out>) at main.cc:64
        base = 0x7f6142229000

----- End of core dump -----
```

**Comment by acepho** [comment]
Added info:

It is confirmed it happens with LOGIN_OB has null UID. LOGIN_OB is fundamental.

In my master.c

`
object connect()
{
	object login_ob;
	mixed err;
	err = catch(login_ob = new(LOGIN_OB));
	if (err)
	{
		write("Come back later.\n");
		write(err);
		destruct(this_object());
	}
	return login_ob;
}
`

and in my login.c

`
void logon()
{ 
	seteuid(getuid()); // It's weird no bug happens here.
}
`

I found them to be identical to testsuite, so it is not lib related. It is all right when the MUD just start. It will only happen after the MUD has been running for some days. And once it happens, it keeps happening, every LOGIN_OB got null UID. I think it's certain there is a bug hiding somewhere in driver. Hope these clues can help you find it.

I will remove getuid() code from creator_file() in my lib so this bug won't hinder me anymore. Don't know if null UID for LOGIN_OB would cause trouble in other place.
